### PR TITLE
[onert] Eliminate copies when subgraph call/return

### PR DESCRIPTION
**Pull request by wateret**
- Remove redundant copies during controlflow op execution
    - `If` - All copies are removed
    - `While` - Some copies are still there, temporary tensors are
      allocated/freed during execution
- Do not handle a primary graph specially
    - Make model input/outputs `IOTensor` rather than `UserTensor`
    - Also a preparation for NNAPI 1.3 support (Any subgraph can be
      primary)

Draft : #4659

ONE-DCO-1.0-Signed-off-by: Hanjoung Lee <hanjoung.lee@samsung.com>

### CODE REVIEW

**Comment by wateret** [comment]
Performance test : https://github.com/Samsung/ONE/pull/4659#issuecomment-716945449

**Comment by wateret** [comment]
@hyunsik-yoon Sorry about such a huge diff, but it is hard to separate. 😢

File: runtime/onert/core/src/backend/controlflow/kernel/WhileLayer.cc [hunk 1]
```diff
@@ -61,159 +61,88 @@ void WhileLayer::run()
...
   };
 
+  std::vector<ITensor *> op_inputs(_input_tensors.begin(), _input_tensors.end());
+  std::vector<ITensor *> op_outputs(_output_tensors.begin(), _output_tensors.end());
+  // Copying body inputs to outputs when the loop body is never executed
+  if (!getResultCond(cond_output_tensor.get()))
+  {
+    PermuteLayer copy_body_inputs_to_op_outputs{op_inputs, op_outputs, _external_context};
+    copy_body_inputs_to_op_outputs.run();
+    // return;
```
Comment (hyunsik-yoon) [inline runtime/onert/core/src/backend/controlflow/kernel/WhileLayer.cc:93]: should this line be uncommented?

File: runtime/onert/core/src/exec/ExecutorBase.cc [hunk 2]
```diff
@@ -73,29 +56,29 @@ void ExecutorBase::execute(const std::vector<backend::ITensor *> &src_tensors,
...
-      const auto orig_input_shape = input_tensor->getShape();
-      const auto changed_input_shape =
-          convertShape(src_tensor->getShape(), src_tensor->layout(), input_tensor->layout());
-      if (orig_input_shape != changed_input_shape)
-      {
-        input_tensor->set_dynamic();
-      }
-    }
+    assert(input_tensor != nullptr);
+    input_tensor->set_dynamic();
```
Comment (hyunsik-yoon) [inline runtime/onert/core/src/exec/ExecutorBase.cc:68]: Previous code call `set_dynamic` only when "src shape != subgraph input shape"
Shouldn't we need such check here?

File: runtime/onert/core/src/exec/ExecutorBase.cc [hunk 3]
```diff
@@ -111,33 +94,32 @@ void ExecutorBase::execute(const IODescription &desc)
...
-    auto *tensor = dynamic_cast<backend::controlflow::UserTensor *>(_output_tensors[i]);
-    assert(tensor);
-    tensor->set_dynamic(); // It can't be resized but shape could change
+    auto tensor = _output_tensors[i];
+
     if (desc.outputs[i] == nullptr)
       throw std::runtime_error{"Output " + std::to_string(i) + "'s buffer is not set."};
-    tensor->setBuffer(static_cast<uint8_t *>(desc.outputs[i]->buffer), desc.outputs[i]->size);
+    tensor->setUserTensor(static_cast<uint8_t *>(desc.outputs[i]->buffer), desc.outputs[i]->size);
+    tensor->set_dynamic(); // It can't be resized but shape could change
```
Comment (hyunsik-yoon) [inline runtime/onert/core/src/exec/ExecutorBase.cc:122]: Could you explain why we need this line all the time?

File: runtime/onert/core/src/backend/controlflow/IOTensor.h [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+ *
+ * A model I/O tensor could be two types.
+ *
+ * 1. @c UserTensor, if it is the primary graph
+ * 2. Any other derivative of @c IPortableTensor from another backend, otherwise
+ *
+ * To support these, this object indirects everything to the actual tensor pointer.
+ * Exceptionally if it is UserTensor, this class creates and manages it.
+ */
+class IOTensor : public IPortableTensor
```
[the remark this answers is not in the document]
Comment (wateret) [reply runtime/onert/core/src/backend/controlflow/IOTensor.h:41]: I'm not sure if I got you right, I think renaming would just resolve the issue. Either `controlflow` -> `builtin`(`builtin::IOTensor`) or the tensor `IOTensor` -> `SubgraphIOTensor`(`controlflow::SubgraphIOTensor`).

> maybe we may need to have controlflow and builtin backend respectively....

I think just renaming `controflow` backend to `builtin` is enough, as mentioned above. Having two different builtin backends(just to have better names) would not be worth it IMHO. 😓 

I think, if `controlflow::IOTensor` sounds confusing, `controlflow::UserTensor` is also confusing. So what I would suggest is **to rename `controlflow` to `builtin`, and that's all**. What do you think about it? 😃

File: runtime/onert/core/src/backend/controlflow/IOTensor.cc [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+  setUserTensor(nullptr, 0);
+}
+
+void IOTensor::setTensor(IPortableTensor *tensor)
+{
+  assert(tensor);
+  assert(tensor != this);
+  // TODO Handle when layout was changed
+  assert(tensor->layout() == _orig_layout); // Changing layout is not considered yet
+  _user_tensor.reset();
```
[the remark this answers is not in the document]
Comment (wateret) [reply runtime/onert/core/src/backend/controlflow/IOTensor.cc:40]: Yes it is correct. As I mentioned offline, a subgraph can be called as both primary(UserTensor) or non-primary, after compilation. So for each run they could call any of `setTensor` or `setUserTensor`.

File: runtime/onert/core/src/exec/ExecutorBase.cc (hunk 3 above)
Comment (wateret) [reply runtime/onert/core/src/exec/ExecutorBase.cc:122]: It was introduced months ago and I forgot 😅 . I will find out why later. I feel like https://github.com/Samsung/ONE/pull/5010/files/738b9d6a9b056e85e81aff2fe1ba77f2e5e3df07#r522001183 is also relavant.

File: runtime/onert/core/src/exec/ExecutorBase.cc (hunk 2 above)
Comment (wateret) [reply runtime/onert/core/src/exec/ExecutorBase.cc:75]: I may have removed it for testing, and forgot. I just put it back.

Comment (wateret) [reply runtime/onert/core/src/exec/ExecutorBase.cc]: Turns out it does not work, need some time to fix!

**Comment by wateret** [comment]
@hyunsik-yoon Please take another look

File: runtime/onert/core/src/backend/controlflow/IOTensor.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (hyunsik-yoon) [reply runtime/onert/core/src/backend/controlflow/IOTensor.h:41]: I agree.

File: runtime/onert/core/src/exec/ExecutorBase.cc (hunk 2 above)
Comment (hyunsik-yoon) [reply runtime/onert/core/src/exec/ExecutorBase.cc]: It's hard to follow which one is fixed. :crying_cat_face:  I guess this fix is done.

File: runtime/onert/core/src/exec/ExecutorBase.cc (hunk 3 above)
Comment (hyunsik-yoon) [reply runtime/onert/core/src/exec/ExecutorBase.cc:128]: Thinking with `WHILE`,
my guess is that... if the all outputs of body_subgraph become dynamic with this line even though some of them are static, probably all inputs of the `WHILE` op will be turned into dynamic, which will make most tensor inside cond and body subgraph dynamic (even though they can be handled in static way)... 
and I am afraid that this is not aligned with the semantics of current dynamic tensor.

**Comment by wateret** [comment]
@hyunsik-yoon I added a commit - https://github.com/Samsung/ONE/pull/5010/commits/f878a1313671a3f106dcab7a5de238ec20bfe00f

File: runtime/onert/core/src/compiler/ExecutorFactory.cc [hunk 6]
```diff
@@ -278,13 +272,9 @@ ExecutorFactory::createLinearExecutor(std::unique_ptr<compiler::LoweredGraph> lo
...
   runTensorRegistration(lowered_graph.get(), order);
 
-  std::vector<backend::ITensor *> input_tensors;
-  std::vector<backend::ITensor *> output_tensors;
-  if (options.is_primary_subgraph)
-  {
-    input_tensors = initializeModelIOTensors(*lowered_graph, lowered_graph->graph().getInputs());
-    output_tensors = initializeModelIOTensors(*lowered_graph, lowered_graph->graph().getOutputs());
-  }
+  initializeModelIOTensors(
```
[the remark this answers is not in the document]
Comment (wateret) [reply runtime/onert/core/src/compiler/ExecutorFactory.cc:275]: Will fix!

File: runtime/onert/core/src/exec/ExecutorBase.cc (hunk 3 above)
Comment (wateret) [reply runtime/onert/core/src/exec/ExecutorBase.cc:128]: Anyways, this code was exist already(a few lines above), so this is beyond the scope of this PR.

File: runtime/onert/core/src/exec/ExecutorBase.cc (hunk 2 above)
Comment (wateret) [reply runtime/onert/core/src/exec/ExecutorBase.cc]: Yes it's back - https://github.com/Samsung/ONE/pull/5010/commits/f878a1313671a3f106dcab7a5de238ec20bfe00f

**Review by hyunsik-yoon — DISMISSED** [review]
LGTM

**Comment by wateret** [comment]
@nnfw-bot test cross-runtime-test-debug
@nnfw-bot test cross-runtime-test-release

**Comment by ragmani** [comment]
@nnfw-bot test cross-runtime-test-release

**Comment by ragmani** [comment]
@nnfw-bot test cross-runtime-test-debug

**Comment by wateret** [comment]
@hyunsik-yoon Rebased the commit due to the CI failure.

**Review by hyunsik-yoon — DISMISSED** [review]
LGTM

**Comment by wateret** [comment]
@hyunsik-yoon Sorry, another rebase was done 😅

**Review by hyunsik-yoon — APPROVED** [review]
LGTM
